src/main/java/org/omnifaces/cdi/BeanStorage.java

```diff
@@ -100,12 +101,13 @@ private static String getBeanId(Contextual<?> type) {
     /**
      * Destroy all beans managed so far.
      */
-    public synchronized void destroyBeans() { // Not sure if synchronization is absolutely necessary. Just to be on safe side.
-        for (Object bean : beans.values()) {
-            destroy(bean);
-        }
-
-        beans.clear();
+    public void destroyBeans() { // Not sure if synchronization is absolutely necessary...
+        // by using ConcurrentHashMap.keySet + remove(),
+        // should be safe to not use synchronization on the method
+        beans.keySet().stream()
+                      .map(beans::remove)
+                      .filter(Objects::nonNull)
+                      .forEach(Beans::destroy);

```

Comment:
Won't work if a bean later in the iteration references in its @PreDestroy a bean which happens to already be removed earlier. Remove must happen after completing the destroy on each bean.